### PR TITLE
feat: support key rotation encryption

### PR DESCRIPTION
**Pull request by debjyoti-pandit**

### CODE REVIEW

File: api/v1/storagecluster_types.go [hunk 1]
```diff
@@ -393,6 +394,18 @@ type EncryptionSpec struct {
...
+	// +optional
+	KMS KMSSecuritySpec `json:"kms,omitempty"`
+}
+
+type KMSSecuritySpec struct {
+	// +optional
+	EnableKeyRotation bool `json:"enableKeyRotation,omitempty"`
+	// +optional
+	Schedule string `json:"schedule,omitempty"`
+}
```
Comment (umangachapagain) [inline api/v1/storagecluster_types.go:407]: We already have `KeyManagementService` just above this. Can't these spec be added there?

**Review by umangachapagain — CHANGES_REQUESTED** [review]

**Review by Rakshith-R — CHANGES_REQUESTED** [review]
/hold

This pr is directly changing noobaa and rook CRDs which it should not.

it needs to wait till related changes are in respective repos. 

cc @SanjalKatiyar

File: controllers/storagecluster/cephobjectstores.go [hunk 2]
```diff
@@ -227,6 +227,8 @@ func (r *StorageClusterReconciler) newCephObjectStoreInstances(initData *ocsv1.S
 			obj.Spec.Security = &cephv1.ObjectStoreSecuritySpec{
 				SecuritySpec: cephv1.SecuritySpec{
 					KeyManagementService: cephv1.KeyManagementServiceSpec{
+						EnableKeyRotation: initData.Spec.Security.KeyManagementService.EnableKeyRotation,
+						Schedule: initData.Spec.Security.KeyManagementService.Schedule,
```
Comment (SanjalKatiyar) [inline controllers/storagecluster/cephobjectstores.go:231]: not needed here... noobaa and ceph are only supported components...

File: controllers/storagecluster/cephobjectstores.go [hunk 3]
```diff
@@ -227,6 +227,8 @@ func (r *StorageClusterReconciler) newCephObjectStoreInstances(initData *ocsv1.S
 			obj.Spec.Security = &cephv1.ObjectStoreSecuritySpec{
 				SecuritySpec: cephv1.SecuritySpec{
 					KeyManagementService: cephv1.KeyManagementServiceSpec{
+						EnableKeyRotation: initData.Spec.Security.KeyManagementService.EnableKeyRotation,
```
Comment (SanjalKatiyar) [inline controllers/storagecluster/cephobjectstores.go:230]: no need to make changes to cephobjectstores... that is out of scope of this epic...

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 4]
```diff
@@ -218,6 +218,10 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 			nb.Spec.Security.KeyManagementService.ConnectionDetails = kmsConfig.Data
 		}
 	}
+	if sc.Spec.Encryption.Enable && sc.Spec.Encryption.ClusterWide && !sc.Spec.Encryption.KeyManagementService.Enable {
+		nb.Spec.Security.KeyManagementService.EnableKeyRotation = sc.Spec.Security.KeyManagementService.EnableKeyRotation
+		nb.Spec.Security.KeyManagementService.Schedule = sc.Spec.Security.KeyManagementService.Schedule
```
Comment (SanjalKatiyar) [inline controllers/storagecluster/noobaa_system_reconciler.go:223]: nb.Spec.Security.KeyManagementService.EnableKeyRotation = sc.Spec.Security.KeyManagementService.EnableKeyRotation (if this field is present in the StorageCluster CR)
true (default value, if this field is missing from the CR)

nb.Spec.Security.KeyManagementService.Schedule = 
sc.Spec.Security.KeyManagementService.Schedule (if this field is present in the StorageCluster CR)
'@weekly' (default value, if this field is missing from the CR)

Comment (SanjalKatiyar) [reply controllers/storagecluster/noobaa_system_reconciler.go:223]: do the same for ceph as well !

Comment (SanjalKatiyar) [reply controllers/storagecluster/noobaa_system_reconciler.go:223]: discussed with @Rakshith-R, will be good to have this for upgrade scenarios (4.12 to 4.13)...

Comment (SanjalKatiyar) [reply controllers/storagecluster/noobaa_system_reconciler.go:223]: basically:
ocs opr should pass on any value which is present in CR...
if none present use "true" and  "weekly" by default...

**Pull request closed without merging**